### ISSUE TITLE
Groups should be able to have multiple parents

### ISSUE THREAD

**Issue by RalfS4M** [keycloak/keycloak#11752]
### Description

Keycloak should support multiple parent groups on groups and users.

### Motivation

We are synching Active Directory users and groups to Keycloak via the `ldap` user federation provider in conjunction with a group mapper of type `group-ldap-mapper`.

Productive AD structures  often (mostly?) have structures with groups having multitple parent groups - whereas Keycloak group only may have a single parent. This causes LDAP sync to fail as long as the **Preserve Group Inheritance** switch in the `group-ldap-mapper` configuration is set.  Disabling this switch results in losing the whole AD structure within Keycloak.

Thus, it would be extremely helpful if Keycloak could be extended to support multiple parent groups.

### Details

First thoughts on the implementation after having had a look on the source code:

- extend database structure to allow multiple parents on a group by adding a n:m mapping table
- adjust API to reflect multiple parents
- adjust user interface to either show the groups structure properly - or at least not to break if a  group has multiple parent; a simplification could be to build the tree shown in the UI only based on the first parent of a group and show the other parents as additional information
- migrate existing data to the new group table structure

**Comment by Markus-Schwer** [comment]
This feature would be really usefull and I think a lot of people have similar use-cases. Would be great to get an update on this.

**Comment by Syniex** [comment]
I am wondering if there any update regarding this, it will be game changer to keep the hierarchical group.

**Comment by tsondergaard** [comment]
I agree that having the group hierarchy beeing a tree rather than a graph is too limiting since LDAP/Active Directory seems like an important use case for KeyCloak and it cannot correctly represent the user group hierachy of LDAP. KeyCloak gets partially around this by flattening LDAP imported groups. This will work for normal user login in applications, but it is a problem when an application implements an administrative interface that actually uses groups to manage data.

See https://issues.redhat.com/browse/KEYCLOAK-7976 for an earlier report of this issue.

**Comment by sangdrax8** [comment]
I was looking into Keycloak in front of my existing OpenLDAP install, and this issue is exactly why I can't deploy it at this time.  Something as basic as having a "admin" group that ends up being in multiple other groups, causes keycloak to report multiple parents and stop working.

It also seems that if I don't retain the structure, it also stops importing any groups that are members of a group.  So I can't even flatten it in keycloak and ensure LDAP has all sub-groups defined as members in the parent group, as it only imports that specific users into the group.

**Comment by salem-med** [comment]
it will be usefull if keycloak supports multiple parents groups

**Comment by philipp1992** [comment]
is there any way to solve this issue: detected to have multiple parents. This is not allowed in Keycloak.

**Comment by jasontseu-bgroup** [comment]
I think it is possible to remove the group inheritance and build virtual parent relationship table to store it. 
But 
1) it will lose another feature which is sync back those relationship to LDAP.
2) need to build graph relationship view in UI.
3) need to build API to expose the graph relationship.

**Comment by fredrikstrange** [comment]
This Keycloak limitation makes it difficult or even impossible to map best-practice AD structure. Please fix.

**Comment by hillout** [comment]
Stumbled upon the same problem..Hopefully KC team will fix this.

**Comment by KajKunstheim** [comment]
> I agree that having the group hierarchy beeing a tree rather than a graph is too limiting since LDAP/Active Directory seems like an important use case for KeyCloak and it cannot correctly represent the user group hierachy of LDAP. KeyCloak gets partially around this by flattening LDAP imported groups. This will work for normal user login in applications, but it is a problem when an application implements an administrative interface that actually uses groups to manage data.
> 
> See https://issues.redhat.com/browse/KEYCLOAK-7976 for an earlier report of this issue.

Looks like devs don't care since 2018

**Comment by shahargl** [comment]
is anybody found a workaround for this?

**Comment by mqu** [comment]
we are also facing this issue with popup error : data could not be synced due to "GroupsMultipleParent".

**Comment by sachingaikwad123** [comment]
I am also looking for this feature. I am not using LDAP sync but I have my own hierarchical groups where multiple-parents are supported for groups. How to sync such groups to Keycloak Groups?

1. Anyone found any workaround for this problem?
2. Thinking out loud: Can we do anything with 'Group Attributes'? Specify additional parents in 'Group Attributes'. In case, multiple parent cases is smaller in number, then this extra overhead of attribute should be OK?

**Comment by adamboutcher** [comment]
+1 for this issue.

**Comment by ciroiriarte** [comment]
would love to see this working, stuck with nested groups!

**Comment by FranciscoTrigo** [comment]
I would also love to get this working

**Comment by kyale** [comment]
+1

**Comment by 53c70r** [comment]
+1

**Comment by mqu** [comment]
+1

**Comment by FranciscoTrigo** [comment]
Anybody know of a workaround?

**Comment by thngz** [comment]
+1

**Comment by WhiteMonsterDark** [comment]
Is there any news on Keycloak's support for multi-parent groups during LDAP synchronisation?

**Comment by ciroiriarte** [comment]
Hello!

After diving into this, I realized it's supported although in a way I was
expecting.

If you create nested groups in your LDAP backend you can still get the
right people in the right group, you just won't see the literal group as a
member of a parent group. Everything is flattened in a single parent group.

Unexpected, but works for me.

Regards,
CI.-

**Comment by WhiteMonsterDark** [comment]
> Hello!
> 
> After diving into this, I realized it's supported although in a way I was
> expecting.
> 
> If you create nested groups in your LDAP backend you can still get the
> right people in the right group, you just won't see the literal group as a
> member of a parent group. Everything is flattened in a single parent group.
> 
> Unexpected, but works for me.
> 
> Regards,
> CI.-
> […](#)

Hello  @ciroiriarte , 
thanks for the feedback.

Right now it might be a solution for me to have all users flattened in the parent group.
However, in my case I cannot get this configuration.
What parameters do I need to configure to achieve this?

**Comment by ciroiriarte** [comment]
fixing the wording:

After diving into this, I realized it's supported although in a way I was NOT expecting.

Play with these parameters to align the setup to what you want/expect:

**preserve.group.inheritance**

If group inheritance will be imported from LDAP. Valid values are "true" and "false".

**user.roles.retrieve.strategy**

"LOAD_GROUPS_BY_MEMBER_ATTRIBUTE": user group membership will be determined by checking the member attribute of groups.

"GET_GROUPS_FROM_USER_MEMBEROF_ATTRIBUTE": The memberOf attribute of the user will determine group membership.

"LOAD_GROUPS_BY_MEMBER_ATTRIBUTE_RECURSIVELY": This value causes Keycloak to load user roles recursively. This method is used only when integrated with Active Directory.

**Comment by patschi** [comment]
I found this issue by wanting to have Keycloak-Groups (also imported from a Microsoft AD) to include all users including any nested groups. I also encountered error "GroupsMultipleParent".

While not entirely matching with issue author's ask, I got a flattened, correctly-recursive resolved users in all groups by adjusting:

**Preserve Group Inheritance**: OFF
**User Groups Retrieve Strategy**: `GET_GROUPS_FROM_USER_MEMBEROF_ATTRIBUTE`
**Member-Of LDAP Attribute**: `memberOf:1.2.840.113556.1.4.1941:`

The LDAP query selector `memberOf:1.2.840.113556.1.4.1941:` is AD-specific and asks AD to essentially search for all groups - including nested - where the user is part of.

**Comment by thngz** [comment]
> I found this issue by wanting to have Keycloak-Groups (also imported from a Microsoft AD) to include all users including any nested groups. I also encountered error "GroupsMultipleParent".
> 
> While not entirely matching with issue author's ask, I got a flattened, correctly-recursive resolved users in all groups by adjusting:
> 
> **Preserve Group Inheritance**: OFF **User Groups Retrieve Strategy**: `GET_GROUPS_FROM_USER_MEMBEROF_ATTRIBUTE` **Member-Of LDAP Attribute**: `memberOf:1.2.840.113556.1.4.1941:`
> 
> The LDAP query selector `memberOf:1.2.840.113556.1.4.1941:` is AD-specific and asks AD to essentially search for all groups - including nested - where the user is part of.

Hello, this displays the groups and its users correctly when viewed in the Groups tab in Keycloak, but navigating onto specific user's profile, the group membership is not displayed (says that user is not part of any group). I need the user and its groups in the Access Token. Any way to do this?

**Comment by labiloute** [comment]
+1 for a full integration of LDAP groups including nested groups
Other workaround does not seems sufficient for us, sorry.

**Comment by thngz** [comment]
> +1 for a full integration of LDAP groups including nested groups Other workaround does not seems sufficient for us, sorry.

This is what got me closer to a working solution, have you tried this?
https://github.com/keycloak/keycloak/issues/26699#issuecomment-3146877259

**Comment by labiloute** [comment]
> > +1 for a full integration of LDAP groups including nested groups Other workaround does not seems sufficient for us, sorry.
> 
> This is what got me closer to a working solution, have you tried this? [#26699 (comment)](https://github.com/keycloak/keycloak/issues/26699#issuecomment-3146877259)

Hi @thngz .
I found this a bit too tinkering for my needs and amount of users.
I am able to wait a bit unless an app have to use one of this parent nested groups, which can happen anytime btw.
I keep your solution in mind if needed, many thanks for suggestion.

**Comment by itcultus** [comment]
+1

**Comment by HagerError404** [comment]
+1

**Comment by maglub** [comment]
We are in an environment where we assign users to a group layer that we call "business roles".

So user A could be member of the groups "Receptionists" and user B is member of "Gardeners".

Both "Receptionists" and "Gardeners" can be member of the technical group (configured in an application), say "app-x-users".

This issue is breaking our (and many others') model.

**Comment by marlenekoh** [comment]
+1, seems to relate to https://github.com/keycloak/keycloak/issues/45291
